dvc/repo/__init__.py

```diff
@@ -356,6 +357,16 @@ def pipelines(self, from_directory=None):
             G.subgraph(c).copy() for c in nx.weakly_connected_components(G)
         ]
 
+    def make_out_dir_filter(self, outs, root_dir):
```

Comment:
This is a microoptimization of a single method. I suggest keeping it inside.

Reply:
BTW, I wonder if this is more of an optimization or a slowdown. The filter is very inefficient, has quadratic `dirs x outs` complexity, and lots of `normpath` calls, which touch fs.

Reply:
But if we leave it in `stages`, the method is long and unreadable. How about performing whole filtering in separate, properly named method?

Reply:
As to performance, I think it is optimization, usually project don't have many directories. On the other hand, data directory for some image processing task (for example) will have extensive amount of files. If we do not filter out the data folder during search for stages, we will add a lot of checks to the main loop.

Reply:
Some people use recursive add and thus thousands of outs, this will make this slow. On method, let's not touch it at least not in this PR. It goes out of control already)